modules/nw-egressnetworkpolicy-about.adoc

```diff
@@ -0,0 +1,70 @@
+// Module included in the following assemblies:
+//
+// * networking/openshift-sdn/configuring-egress-firewall.adoc
+
+[id="nw-egressnetworkpolicy-about_{context}"]
+= How an egress firewall works in a project
+
+As an {product-title} cluster administrator, you can use an _egress firewall_ to
+limit the external IP addresses that some or all Pods can access from within the
+cluster. Egress firewall policy supports the following scenarios:
+
+- A Pod can only connect to internal hosts, and cannot initiate connections to
+the public Internet.
+- A Pod can only connect to the public Internet, and cannot initiate connections
+to internal hosts that are outside the {product-title} cluster.
+- A Pod cannot reach specified internal subnets or hosts outside the {product-title} cluster that should be unreachable.
+
+An egress firewall policy is configured by creating an EgressNetworkPolicy Custom Resource (CR) object and specifying an IP address range in CIDR format or by specifying a DNS name.
+For example, you can allow one project access to a specified IP range but deny the same access to a different project, or you can restrict application developers from updating from Python pip mirrors, and force updates to come only from approved sources.
+
+[IMPORTANT]
+====
+You must have OpenShift SDN configured to use either the network policy or multitenant modes to configure egress firewall policy.
+
+If you are using network policy mode, egress policy is compatible with only one policy per project, and will not work with projects that share a network, such as global projects.
+====
+
+[CAUTION]
+====
+Exposing services by creating routes will ignore egress network policies.
```

Comment:
"by creating routes" with custom, external, endpoints "will ignore ..."

The trick is that you could make a service without a selector (https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors) and then create endpoint objects for it with external IP addresses.

Then you hit the service, and, because if the iptables rules kube-proxy installs, the kernel magically sends the traffic to the external resource, bypassing the egress firewall.

The advice is to disable creation of routes (or perhaps endpoint objects instead, if that's possible) where that's a concern.

Reply:
"routes" here refers to `Route` objects. You can't actually hack around ENP with selectorless services because we (openshift-sdn) control kube-proxy and can stop it from creating iptables rules for any service that would allow bypassing ENP. However, we don't control the routers, so if you make a `Route` that points to an external IP, you can use that bypass ENP.

Reply:
I'll need to find a way to reword this, so it makes more sense. As written, I'm not sure what we were trying to express.

Reply:
@danwinship, how about:

> Egress network policy rules do not apply to service endpoint traffic. An Ingress Controller handles traffic for a service, bypassing the egress firewall.

Reply:
No, ENPs *do* apply to service endpoints. They don't apply to *Routes*.

Egress network policy rules do not apply to traffic that goes through Routers. A user who can create `Route`s can bypass Egress network policy rules by creating a route that points to a forbidden destination.

Reply:
@danwinship, great, thanks!